### ISSUE TITLE
dispatchReturn with reject() in a validate hook dispatches a success type action not a fail type

### ISSUE THREAD

**Issue by DenisNeustroev** [jeffbski/redux-logic#52]
When turn on dispatchReturn in processOptions and after reject action in a validate hook, action with successType will be dispatched [demo on jsfiddle](https://jsfiddle.net/g1hmrjhv/2/), pay attention to lines 30 and 96.

**Comment by TraceyK14** [comment]
No response in 8 months? @jeffbski 
I am having a similar problem, I am trying to follow your form validation example but slightly tweaked as I don't wanna store the fields in state.

It gets to the reject part but doesn't call the 'loginFieldsInvalid' action and then continues through to process. If I don't pass an action to reject it will work but I need the errors object so that doesn't help much haha

Any guidance would be appreciated.

```
import { createLogic } from 'redux-logic';
import { LOGIN_USER, LOGIN_SUCCESSFUL, LOGIN_FAILED, loginFieldsInvalid } from './actions';

const validateFields = fields => {
    const errors = []
    if(!fields.email) { errors.push('Email is required') }
    if(!fields.password) { errors.push('Password is required') }
    return errors
}

const userLoginLogic = createLogic({
    type: LOGIN_USER,
    validate({ action }, allow, reject) {
        const errors = validateFields(action.payload)
        if(!errors.length){
           return allow(action)
        } 
        reject(loginFieldsInvalid(errors))
        
    },
    processOptions: {
        successType: LOGIN_SUCCESSFUL,
        failType: LOGIN_FAILED
    },
    process({ http, action }) {
        let loginRequest = http.post('/users/login', action.payload)
            .then(response => console.log(response.data))
            .catch(err => alert(err))

        return loginRequest
    }
})

export default [
    userLoginLogic
]
```

**Comment by jeffbski** [comment]
Sorry @TraceyK14 and @DenisNeustroev  I will take a look this weekend and respond back.

**Comment by TraceyK14** [comment]
@jeffbski awesome, thank you! If you need anymore information from me please let me know.

**Comment by jeffbski** [comment]
I tried to run the jsfiddle from @DenisNeustroev but it doesn't seem to run right now.

@TraceyK14 I am looking at what you have posted and it all looks good. I don't see the code for loginFieldsInvalid, so I am wondering if there is a problem in there. 

It looks like you are rejecting with an action and that will be dispatched and process won't run. 

I went back and ran my form-validation example to make sure it was still working and it seems to be dispatching the invalid action and showing that as validation errors. I added a process to my example to see that validate reject was indeed stopping process from being called and it was.

Would it be possible for you to show me what loginFieldsInvalid and even better could you create a jsfiddle or equivalent that demonstrates the problem? I could debug through it. Right now I can only speculate on what is going on since the code I see looks fine.

**Comment by TraceyK14** [comment]
@jeffbski Here is the code in my actions.js which contains the loginFieldsInvalid action.
Below is the link to a working fiddle and the reject action is still not firing. The isLoggedIn value should never be true. I have the button fire the login action with empty credentials and I also have the errors array being console.logged so you can see that it is correctly identifying the errors and storing them, it also doesn't enter the if statement so the "allow(action)" is never called yet it still continues.

 [JSFIDDLE LINK](https://jsfiddle.net/cdpu11ys/2/)

If you need anything else please let me know

```
export const key = 'login';

export const LOGIN_USER = 'LOGIN_USER';
export const LOGIN_SUCCESSFUL = 'LOGIN_SUCCESSFUL';
export const LOGIN_FAILED = 'LOGIN_FAILED';
export const LOGIN_FIELDS_INVALID = 'LOGIN_FIELDS_INVALID';

export const actionTypes = {
    LOGIN_USER,
    LOGIN_SUCCESSFUL,
    LOGIN_FAILED,
    LOGIN_FIELDS_INVALID
}

export const loginUser = data => ({ type: LOGIN_USER, payload: data });
export const loginFieldsInvalid = data => ({ type: LOGIN_FIELDS_INVALID, payload: data })

export const actions = {
    loginUser,
    loginFieldsInvalid
}
```

**Comment by jeffbski** [comment]
@TraceyK14 Thanks! I see the issue now and am working on a fix. It has to with some refactoring I did to clean things up. It is not calling the process fn on the failed validation but what is going wrong is that the reject action is being wrapped with the successType when that should only happen for actions dispatched from the process function. 

To work around the issue right now, don't use the processOptions.successType if you are rejecting with a new action type. Instead just decorate the success value in your promise until this is fixed.

```js
const userLoginLogic = createLogic({
    type: LOGIN_USER,
    validate({ action }, allow, reject) {
        const errors = validateFields(action.payload)
        console.log(errors)
        if(!errors.length){
           return allow(action)
        } 
        reject(loginFieldsInvalid(errors))
        
    },
    processOptions: {
        // successType: LOGIN_SUCCESSFUL,  // wrapping in process for now as work around
        failType: LOGIN_FAILED
    },
    process({ http, action }) {
        let request = http.post('/users/login', action.payload )
            .then(response => ({ type: LOGIN_SUCCESSFUL, payload: response }));
        
        return request

    }
})
```


I'm sorry I didn't catch this sooner. I'll have the proper tests and fix ready soon.

**Comment by TraceyK14** [comment]
@jeffbski I'm glad you were able to figure it out! And thank you for the workaround, it's working correctly now. 

No worries, I'm just glad I was able to help.

**Comment by jeffbski** [comment]
original defect is fixed in #103

**Comment by jeffbski** [comment]
Deployed in v0.12.4.

Thanks again @TraceyK14 for helping me find this.
